src/generate-types.ts

```diff
@@ -261,1117 +79,1403 @@ const statusCodes = [
   504
 ];
 
-function expandedWildcardCode(code: string): number[] {
-  if (code === 'default') {
-    return [];
-  }
-  const wildcard = new RegExp(code.replace(/X/g, '.'));
-  return statusCodes.filter(code => wildcard.test('' + code));
+// list of files for which we have an action to generate it already
+// to prevent calling an action to generate it again and causing maybe some race conditions
+const generatedFiles: Set<string> = new Set();
+
+// query to figure out whether the file is being already generated
+function isGenerating(file: string): boolean {
+  return generatedFiles.has(file);
 }
 
-function generateResponseType(opt: Options, op: string, responses: oas.ResponsesObject) {
-  if (!responses) {
-    return assert.fail('missing responses');
+export function run(options: Options) {
+  options.targetFile = './' + path.normalize(options.targetFile);
+  if (isGenerating(options.targetFile) && !options.forceGenerateTypes) {
+    return;
   }
-  const allSpecifiedStatuses: number[] = [];
-  Object.keys(responses)
-    .map(expandedWildcardCode)
-    .map(codes =>
-      codes.map(code => {
-        assert(allSpecifiedStatuses.indexOf(code) < 0, 'duplicate status code ' + code);
-        allSpecifiedStatuses.push(code);
-      })
+  generatedFiles.add(options.targetFile);
+  deprecated(
+    options.externalOpenApiImports.length > 0,
+    "'externalOpenApiImports' is deprecated. Consider using 'resolve' instead"
+  );
+  deprecated(
+    options.externalOpenApiSpecs,
+    "'externalOpenApiSpecs' is deprecated. Consider using 'resolve' instead"
+  );
+
+  const state = {
+    cwd: path.dirname(options.sourceFile),
+    imports: {} as Record<string, string>,
+    actions: [] as Array<() => Promise<void>>
+  };
+  const builtins = generateBuiltins();
+  const types = generateComponents(options);
+  const queryTypes = generateQueryTypes(options);
+
+  const externals = generateExternals(
+    Object.entries(state.imports).map(([importAs, importFile]) => ({
+      importAs,
+      importFile: resolveModule(options.targetFile, importFile)
+    }))
+  );
+  state.actions.forEach(action => action());
+
+  const sourceFile: ts.SourceFile = ts.createSourceFile(
+    'test.ts',
```

Comment:
what's `test.ts`?

Reply:
it's the filename we lie we are writing to when generating the source text. It does not get written anywhere though. I imagine that it is used for error diagnostics or something but 🤷‍♂️